plugins/event_notification/providers/email/lib/batch/KDispatchEmailNotificationEngine.class.php

```diff
@@ -274,32 +275,34 @@ protected function sendEmail(KalturaEmailNotificationTemplate $emailNotification
 		{
 		    if($recipientsBcc)
 		    {
-			$recipients =  array_slice($recipientsBcc, $recipientsBccHandledCounter, $recipientsBccBulk - 1);
-			foreach ($recipients as $email=>$name)
-			{
-			    $recipientsBccHandledCounter++;
-			    if (filter_var($email, FILTER_VALIDATE_EMAIL))
-			    {
-				KalturaLog::info("Adding recipient to BCC recipients $name<$email> , Index:$recipientsBccHandledCounter");
-				self::$mailer->AddBCC($email, $name);
-			    }
-			}
+				$recipients =  array_slice($recipientsBcc, $recipientsBccHandledCounter, $recipientsBccBulk - 1);
+				foreach ($recipients as $email=>$name)
+				{
+			    	$recipientsBccHandledCounter++;
+			    	if(filter_var($email, FILTER_VALIDATE_EMAIL))
+			    	{
+					KalturaLog::info("Adding recipient to BCC recipients $name<$email> , Index:$recipientsBccHandledCounter");
+					self::$mailer->AddBCC($email, $name);
+			    	}
+				}
 		    }
 
 		    try 
 		    {
-			KalturaLog::info('Sending Bulk');
-			$success = $this::$mailer->Send();
-			if (!$success)
-			{	
-			    throw new kTemporaryException("Sending mail failed: " . $this::$mailer->ErrorInfo);
-			}
-		    } catch (Exception $e) 
+				KalturaLog::info('Sending Bulk');
+				$success = $this::$mailer->Send();
+				if (!$success)
+				{
+			    	throw new kTemporaryException("Sending mail failed: " . $this::$mailer->ErrorInfo);
+				}
+		    }
+			catch (Exception $e)
 		    {
-			throw new kTemporaryException("Sending mail failed with exception: " . $e->getMessage(), $e->getCode());
+				throw new kTemporaryException("Sending mail failed with exception: " . $e->getMessage(), $e->getCode());
 		    }
 
 		    self::$mailer->ClearBCCs();
+			$recipientsBccHandledCounter++;
```

Comment:
indentation
we increase the counter in line 281, why we need it here as well?

Reply:
In each while loop we iterate over 500 users, we slice $recipientsBcc array from index 0 to 499 (== 500 users).
After iterating over all the current bulk of users the value of $recipientsBccHandledCounter is 499.
**Now we have two cases:**
**1.** This is the last iteration, we enter this line of checking if $recipientsBccHandledCounter < count($recipientsBcc),
in case we have only 500 users for example we check if 499 < 500, we'll get True and enter the foreach loop again even we do not have more users.
**2.** We continue to the next iteration, the next bulk of 500, but first of all we slice the $recipientsBcc array from index 499, 500 users. we alse send the email for user 499 in the previous iteration, we have to start now from user 500.